lang/shared/src/main/scala/com/wavesplatform/lang/v1/compiler/ExpressionCompiler.scala

```diff
@@ -259,10 +260,15 @@ object ExpressionCompiler {
         case _      => None
       }
       matchTypes <- cases.traverse(c => handleCompositeType(p, c.caseType, Some(exprTypes), allowShadowVarName))
+      defaultType = exprTypes match {
+        case ANY => ANY
+        case UNION(tl, _) => UNION(tl.filter(t => !matchTypes.contains(t)), None)
+        case _ => NOTHING
```

Comment:
What if union is empty? And what about backward compatibility for v1,2,3?

Reply:
If UNION is empty default case not available.

Reply:
I tried, it is compatible.